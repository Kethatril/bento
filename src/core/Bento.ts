import { resolve } from "path";
import { Loader, Condition, Configuration, Resolve } from "webpack";
import { selector } from "~/utils/env";
import { instantiate } from "~/utils/lang";
import * as resolveModule from "resolve";

export interface BaseConfig {
    homeDir: string;
    outputDir: string;
    entry: {};
    publicPath?: string;
}

export interface RuleMap {
    [ext: string]: Loader | Loader[];
}

export interface Rule {
    include: Condition[];
    use: Loader | Loader[];
}

export interface RuleDescriptor {
    ext: string[];
    rule: Rule;
}

export interface PluginMap {
    [name: string]: any[];
}

export interface PluginDescriptor {
    name: string;
    args: any[];
}

export interface Features {
    sourceMap: boolean;
}

/**
 * Main class of the library
 *
 * @export
 * @class Bento
 */
export default class Bento {
    /**
     * @private
     * @type {BaseConfig}
     * @memberof Bento
     */
    private config: BaseConfig;

    /**
     * @private
     * @type {string}
     * @memberof Bento
     */
    private cwd: string;

    /**
     * @private
     * @type {string}
     * @memberof Bento
     */
    private context: string;

    /**
     * @private
     * @type {RuleDescriptor[]}
     * @memberof Bento
     */
    private rules: RuleDescriptor[] = [];

    /**
     * @private
     * @type {PluginDescriptor[]}
     * @memberof Bento
     */
    private plugins: PluginDescriptor[] = [];

    /**
     * @private
     * @type {Features}
     * @memberof Bento
     */
    private features: Features = {
        sourceMap: true
    };

    /**
     * @private
     * @memberof Bento
     */
    private resolve: (path: string) => string;

    /**
     * @private
     * @memberof Bento
     */
    private select: (choice: {}) => any;

    /**
     * Convenient method to create an instance of Bento
     *
     * @static
     * @param {BaseConfig} config
     * @param {string} [env]
     * @param {string} [cwd]
     * @returns {Bento}
     * @memberof Bento
     */
    public static create(
        config: BaseConfig,
        env?: string,
        context?: string
    ): Bento {
        return new Bento(config, env, context);
    }

    /**
     * Creates an instance of Bento.
     * @param {BaseConfig} config
     * @param {string} [env]
     * @param {string} [cwd]
     * @memberof Bento
     */
    public constructor(
        config: BaseConfig,
        env: string = process.env.NODE_ENV,
        cwd: string = process.cwd()
    ) {
        this.config = config;
        this.cwd = cwd;
        this.context = this.cwd;
        this.resolve = resolve.bind(this, this.context);
        this.select = selector(env);
    }

    /**
     * Create a new RuleDescriptor by a given data and add it to rule list
     *
     * @param {string} ext  File extension
     * @param {Loader | Loader[]} loaders Loaders to use
     * @returns {this}
     * @memberof Bento
     */
    addRule(ext: string | string[], loaders: Loader | Loader[]): this {
        if (typeof ext === "string") ext = [ext];
        this.rules.push({
            ext,
            rule: {
                include: [this.resolve(this.config.homeDir)],
                use: loaders
            }
        });

        return this;
    }

    /**
     * Convenient method to add multiple rules
     *
     * @param {RuleMap} map Map of file extensions and their creators
     * @returns {this}
     * @memberof Bento
     */
    addRules(map: RuleMap = {}): this {
        for (let p in map) {
            this.addRule(p.replace(/\s+/g, "").split(","), map[p]);
        }

        return this;
    }

    /**
     * Create a new PluginDescriptor by given data and add it to plugin list
     *
     * @param {string} name Plugin module name to resolve
     * @param {any[]} args Plugin constructor arguments
     * @returns {this}
     * @memberof Bento
     */
    addPlugin(name: string, args: any[] = []): this {
        this.plugins.push({
            name,
            args
        });

        return this;
    }

    /**
     * Convenient method to add multiple plugins
     *
     * @param {PluginMap} map
     * @returns {this}
     * @memberof Bento
     */
    addPlugins(map: PluginMap = {}): this {
        for (let p in map) {
            this.addPlugin(p, map[p]);
        }

        return this;
    }

    /**
     * Export information to a standard consumable webpack config file
     *
     * @param {(manifest: {}) => void} [override]
     * @returns {Promise<Configuration>}
     * @memberof Bento
     */
    export(override?: (manifest: {}) => void): Promise<Configuration> {
        const description = {
            rules: this.rules,
            plugins: this.plugins
        };

        if (override) override(description);

        const config = {
            name: "bento",
            context: this.context,
            entry: this.config.entry,
            output: {
                path: this.resolve(this.config.outputDir),
                pathinfo: this.select({
                    default: true,
                    production: false
                }),
                filename: this.select({
                    default: "[name].js",
                    production: "[name].[chunkhash].js"
                }),
                publicPath: this.config.publicPath || "/"
            },
            devtool: this.select({
                default: this.features.sourceMap
                    ? "cheap-module-eval-source-map"
                    : null,
                production: this.features.sourceMap ? "source-map" : false
            }),
            resolve: {
                alias: {
                    "~": this.resolve(this.config.homeDir)
                }
            },
            module: {
                rules: this.rules.map(desc => {
                    return Object.assign({}, desc.rule, {
                        test: new RegExp(`\\.(${desc.ext.join("|")})$`)
                    });
                })
            },
            plugins: this.plugins.map(desc => {
                const module = resolveModule.sync(desc.name, {
                    basedir: this.context
                });

                const Plugin = require(module);
                return instantiate(Plugin, desc.args);
            })
        };

        return Promise.resolve(config);
    }

    /**
     * Enable / disable available features
     *
     * @param {string} flag
     * @param {boolean} value
     * @returns {this}
     * @memberof Bento
     */
    set(flag: string, value: boolean): this {
        this.features[flag] = value;
        return this;
    }
}
