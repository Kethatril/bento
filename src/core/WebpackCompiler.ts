import WebpackError from "webpack-promise/js/WebpackError";
import * as webpack from "webpack";
import { Compiler } from "webpack";

/**
 * Simple wraper for `webpack`
 *
 * @export
 * @class WebpackCompiler
 */
export default class WebpackCompiler {
    /**
     * @public
     * @type {webpack.Compiler.Watching | webpack.Compiler}
     * @memberof WebpackCompiler
     */
    public webpack: webpack.Compiler.Watching | webpack.Compiler;

    /**
     * Execute webpack
     *
     * @param {Configuration} config
     * @returns {Promise<Stats>}
     * @memberof WebpackCompiler
     */
    compile(config: webpack.Configuration): Promise<webpack.Stats> {
        return new Promise((resolve, reject) => {
            this.webpack = webpack(config, (err, stats) => {
                if (err || stats.hasErrors() || stats.hasWarnings())
                    reject(new WebpackError(err, stats));
                else resolve(stats);
            });
        });
    }
}
