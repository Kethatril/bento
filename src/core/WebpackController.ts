import * as webpack from "webpack";
import { Compiler, Configuration, Stats } from "webpack";
import WebpackCompilationError from "~/errors/WebpackCompilationError";

/**
 * Simple controller for `webpack`
 *
 * @export
 * @class WebpackController
 */
export default class WebpackController {
    /**
     * @public
     * @type {Compiler.Watching | Compiler}
     * @memberof WebpackController
     */
    public compiler: Compiler.Watching | Compiler;

    /**
     * @type {Configuration}
     * @memberof WebpackController
     */
    public config: Configuration;

    /**
     * @type {Stats}
     * @memberof WebpackController
     */
    public stats: Stats;

    /**
     * Creates an instance of WebpackController.
     *
     * @param {Configuration} config
     * @memberof WebpackController
     */
    constructor(config: Configuration) {
        this.config = config;
    }

    /**
     * Execute webpack
     *
     * @returns {Compiler.Watching | Compiler}
     * @memberof WebpackController
     */
    compile(): Promise<Stats> {
        return new Promise((resolve, reject) => {
            this.compiler = webpack(this.config, (err, stats) => {
                if (err) return reject(err);
                if (stats.hasErrors())
                    return reject(new WebpackCompilationError(stats));
                if (stats.hasWarnings()) {
                    console.warn(stats.toJson().warnings);
                }
                resolve((this.stats = stats));
            });
        });
    }
}
