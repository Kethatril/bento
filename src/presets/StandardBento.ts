import {
    optimize,
    DefinePlugin,
    Loader,
    NamedModulesPlugin,
    NamedChunksPlugin
} from "webpack";
import { extract } from "extract-text-webpack-plugin";
import { basename, extname } from "path";
import Bento, { Features as BaseFeatures, Env } from "../core/Bento";
import { selector } from "../utils/lang";

export interface Features extends BaseFeatures {
    extractCss: boolean;
    writeManifest: boolean;
}

export default class StandardBento extends Bento {
    /**
     * @protected
     * @type {Features}
     * @memberof StandardBento
     */
    protected features: Features;

    /**
     * @see Bento#configure
     * @protected
     * @memberof StandardBento
     */
    protected configure() {
        // set default addtional features flag
        this.set("extractCss", true).set("writeManifest", true);

        // SCRIPT
        this.addRule(["js", "jsx"], {
            loader: "babel-loader",
            options: {
                presets: [
                    [
                        "@babel/preset-env",
                        {
                            useBuiltIns: "usage"
                        }
                    ]
                ]
            }
        })
            .addPlugin("clean-webpack-plugin", [
                this.config.outputDir,
                {
                    root: this.cwd
                }
            ])
            .addPlugin("webpack-manifest-plugin", (env?: string): any[] => {
                return [
                    {
                        writeToFileEmit: this.features.writeManifest
                    }
                ];
            })
            .addPlugin(NamedChunksPlugin, [
                chunk => {
                    if (chunk.name) {
                        return chunk.name;
                    }

                    return chunk.mapModules(m => {
                        return basename(m.request, extname(m.request));
                    });
                }
            ])
            .addPlugin(NamedModulesPlugin, [])
            .addPlugin(
                optimize.CommonsChunkPlugin,
                [
                    {
                        name: "vendor",
                        minChunks(module) {
                            return /node_modules/.test(module.resource);
                        }
                    }
                ],
                Env.Production
            )
            .addPlugin(
                optimize.CommonsChunkPlugin,
                [
                    {
                        name: "runtime"
                    }
                ],
                Env.Production
            )
            .addPlugin(DefinePlugin, (env?: Env): any[] => {
                return [
                    {
                        "process.env": {
                            NODE_ENV: JSON.stringify(env)
                        }
                    }
                ];
            })

            .addPlugin("uglifyjs-webpack-plugin", [], Env.Production);

        // STYLE
        this.addRule("scss", (env?: string): Loader[] => {
            return extract({
                fallback: {
                    loader: "style-loader",
                    options: {
                        sourceMap: this.features.sourceMap
                    }
                },
                use: [
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap,
                            minimize:
                                env === Env.Production
                                    ? { preset: "default" }
                                    : false
                        }
                    },
                    {
                        loader: "postcss-loader",
                        options: {
                            ident: "postcss",
                            plugins() {
                                const autoprefixer = require("autoprefixer");
                                return [autoprefixer()];
                            },
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "resolve-url-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "sass-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ]
            });
        })
            .addRule(
                "css",
                [
                    {
                        loader: "style-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ],
                /node_modules/
            )
            .addPlugin("extract-text-webpack-plugin", (env?: string): any[] => {
                const select = selector(env);

                return [
                    {
                        filename: select({
                            default: "[name].css",
                            production: "[name].[contenthash:8].css"
                        }),
                        disable: !(
                            this.features.extractCss && env === Env.Production
                        )
                    }
                ];
            });

        // FONTS & MEDIA
        this.addRule(
            [
                "ttf",
                "eot",
                "woff",
                "woff2",
                "jpeg",
                "jpg",
                "png",
                "gif",
                "svg",
                "mp4",
                "webm"
            ],
            (env?: string): Loader => {
                const select = selector(env);

                return {
                    loader: "file-loader",
                    options: {
                        name: select({
                            default: "[name].[ext]",
                            production: "[name].[hash:8].[ext]"
                        })
                    }
                };
            },
            [this.homeDir, /node_modules/]
        );
    }
}
