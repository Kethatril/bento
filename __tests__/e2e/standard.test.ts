import StandardBento from "~/presets/StandardBento";
import WebpackCompiler from "~/core/WebpackCompiler";
import { Env } from "~/core/Bento";

describe("build", () => {
    const build = function(entry, env?: Env) {
        const bento = StandardBento.create({
            homeDir: "./__tests__",
            outputDir: "/"
        }).bundle("main", `~/fixtures/${entry}.js`);

        const compiler = new WebpackCompiler();
        const files = {};

        return bento.export(env).then(config => {
            config.devtool = false;
            const promise = compiler.compile(config);

            compiler.webpack.outputFileSystem.writeFile = function(
                name,
                content,
                callback
            ) {
                files[name] = content.toString("utf-8");
                callback();
            };

            return promise.then(stats => {
                return [files, stats];
            });
        });
    };

    it("exports a correct `config` for webpack", () => {
        return build("main2").then(([files, stats]) => {
            expect(Object.keys(files)).toEqual([
                "/RobotoMono-Regular.ttf",
                "/giphy.gif",
                "/cat.gif",
                "/nice.jpg",
                "/main.js"
            ]);
            const bundle = files["/main.js"];

            expect(bundle).toContain("function __webpack_require__(");
            expect(bundle).toContain("__webpack_require__(/*! ~/fixtures/a */");
            expect(bundle).toContain("__webpack_require__(/*! ./style.scss */");
            expect(bundle).toContain("./__tests__/fixtures/main2.js");
            expect(bundle).toContain("./__tests__/fixtures/a.js");
            expect(bundle).toContain("./__tests__/fixtures/b.js");
            expect(bundle).toContain(
                "./__tests__/fixtures/node_modules/m1/a.js"
            );
            expect(bundle).toContain(
                "./__tests__/fixtures/node_modules/m1/s.css"
            );
            expect(bundle).toContain("This is a");
            expect(bundle).toContain("This is b");
            expect(bundle).toContain("This is m1/a");
            expect(bundle).toContain("var x = 1;");
            expect(bundle).toContain("color: red;");
            expect(bundle).toContain("font-size: 12px;");
            expect(bundle).toContain("console.log(undefined);");

            expect(bundle).not.toContain("4: function(");
            // expect(bundle).not.toContain("window");
            expect(bundle).not.toContain("jsonp");
        });
    });

    // it.only("exports a correct `config` for webpack in production", () => {
    //     return build("main2", "production").then(([files, stats]) => {
    //         console.log(Object.keys(files));

    //         console.log(files["/main.e9e46029.css"]);
    //     });
    // });
});
