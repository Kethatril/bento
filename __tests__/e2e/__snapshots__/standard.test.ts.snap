// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build exports a correct \`config\` for webpack 1`] = `
Object {
  "RobotoMono-Regular.ttf": "/RobotoMono-Regular.ttf",
  "cat.gif": "/cat.gif",
  "giphy.gif": "/giphy.gif",
  "main.js": "/main.js",
  "nice.jpg": "/nice.jpg",
}
`;

exports[`build exports a correct \`config\` for webpack in production 1`] = `
Object {
  "RobotoMono-Regular.ttf": "/RobotoMono-Regular.a48ac416.ttf",
  "cat.gif": "/cat.5082946a.gif",
  "giphy.gif": "/giphy.a47e713b.gif",
  "main.css": "/main.bd753fff.css",
  "main.js": "/main.d8732cfc.js",
  "nice.jpg": "/nice.9c3c4150.jpg",
  "runtime.js": "/runtime.631fbc64.js",
  "vendor.js": "/vendor.a9d97159.js",
}
`;
